modules/mesh-layers/src/scenegraph-layer/scenegraph-layer.js

```diff
@@ -91,8 +86,8 @@ const defaultProps = {
   getPitch: {type: 'accessor', value: x => x.pitch || 0},
   getRoll: {type: 'accessor', value: x => x.roll || 0},
   getScale: {type: 'accessor', value: x => x.scale || [1, 1, 1]},
-  getTranslation: {type: 'accessor', value: x => x.translate || [0, 0, 0]},
-  getMatrix: {type: 'accessor', value: x => x.matrix || null}
+  getTranslation: {type: 'accessor', value: x => x.translation || [0, 0, 0]},
+  getTransformMatrix: {type: 'accessor', value: x => x.transformMatrix || null}
```

Comment:
I am not entirely happy with this hybrid API (multiple accessors go into one attribute). This removes the chance to use constant accessors (i.e. `getScale: [1, 1, 1]` and `getTranslation: [0, 0, 0]`). If I have a large number of instances and I need to change a small portion of the transform (e.g. cars animated by position and heading), it is now a much more expensive operation (recalculating mat4 vs float).

Reply:
That would just push the creation of the matrix to the vertex shader, where it would happen per-vertex instead of per-instance, and per-frame instead of per-update. That will be a performance loss in most cases.

A potential win, if we feel the translation part isn't useful, would be send a mat3 instead of a mat4. That would also kill off the empty 4th row of the the mat4 that's being sent in the current implementation.

Reply:
I like this idea. Additional thoughts:
- Merge `getPitch`, `getRoll`, `getYaw` into `getOrientation` to match `getScale`
- Allow `getOrientation`, `getScale` and `getTransformMatrix` to be constants.
- math.gl's `Quaternion` and `Euler` classes should provide methods that produce mat3 transform matrices

Reply:
Seems like this was already addressed over the weekend, so I am late to the party.

FWIW I will add my 2 cents:

The concerns raised are valid and needed to be addressed. That said, I personally still think we should support 4x4 matrices, at least in `ScenegraphLayer` so that we offer the most flexible integration with arbitrary approaches to generating the transforms.

I believe that we could support both cases efficiently by:

* only having a single `mat4` attribute
* if `getMatrix` is provided, we use that populate the attributes.
* If not, we storing `getScale()`, `getOrientation()`, and `[roll, pitch, yaw]` in the first three columns. From the JS side, each column is an individual attributes and can be set to constants etc.
* If the "component version" is used, we set the forth column to `[0, 0, 0, 0]`, and we can use the value of the last element (`mat[3][3]`) to decide:
* `mat[3][3] == 1`: Matrix can be used directly
* `mat[3][3] == 0`: Matrix needs to be built from col 1-3.

Perf:
* Case 1: (4x4 matrix) is optimally efficient, matrix can be used directly.
* Case 2: (components) The vertex shader perf hit in is outweighed by the JS savings and ability to support constants.